vespalib/src/vespa/vespalib/util/invokeserviceimpl.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "invokeserviceimpl.h"
+#include <cassert>
+
+namespace vespalib {
+
+InvokeServiceImpl::InvokeServiceImpl(duration napTime)
+    : _naptime(napTime),
+      _lock(),
+      _closed(false),
+      _toWakeup(),
+      _thread()
+{
+}
+
+InvokeServiceImpl::~InvokeServiceImpl()
+{
+    {
+        std::lock_guard guard(_lock);
+        assert(_toWakeup.empty());
+        _closed = true;
+    }
+    if (_thread) {
+        _thread->join();
+    }
+}
+
+class InvokeServiceImpl::Registration : public IDestructorCallback {
+public:
+    Registration(InvokeServiceImpl * service, VoidFunc func) noexcept
+        : _service(service),
+          _func(func)
+    { }
+    Registration(const Registration &) = delete;
+    Registration & operator=(const Registration &) = delete;
+    ~Registration() override{
+        _service->unregister(_func);
+    }
+private:
+    InvokeServiceImpl * _service;
+    VoidFunc        _func;
+};
+
+std::unique_ptr<IDestructorCallback>
+InvokeServiceImpl::registerInvoke(VoidFunc func) {
+    std::lock_guard guard(_lock);
+    _toWakeup.push_back(func);
+    if ( ! _thread) {
+        _thread = std::make_unique<std::thread>(InvokeServiceImpl::run, this);
```

Comment:
We do not need a static function here. Since the syntax to call a member is a bit magic, I suggest using a lambda:
`std::make_unique<std::thread>([this]{ runLoop(); })`

Reply:
Fixed